Add Philips air purifier integration

mfrister opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This adds an integration for Philips Air Purifiers, currently tested with a Philips AC2889/10.

The integration currently supports:

* Configuration via config flow
* Device with a single fan entity 
    * On/off
    * Set manual fan speed via percentage (mapped to 1/2/3/turbo)
    * Set preset modes (auto, sleep, allergen, bacteria/virus)
    * Detecting whether the purifier is available (with some delay, related to limitations with device communication). 

Limitations:

* Fan only for now, as requested in the dev docs (only one new platform). I might later add a sensor as well, the code was written with that in mind (connection to the device is managed outside the fan entity).
* No explicit support for multiple purifier models. From what I've seen from existing libraries and custom integrations, they have similar interfaces, but not all models support all features. This MR supports a pretty basic set of features, but I don't know whether all Philips Air Purifiers support them.
* I've forked an existing library for controlling Philips air purifiers ([aioairctrl](https://github.com/betaboon/aioairctrl)), renamed it and published it on PyPI as [phipsair](https://pypi.org/project/phipsair/).
    - The library contained two monkey patches for `aiocoap`, which already seems to already be used at least by `pytradfri`, which in turn is used in Home Assistant as well. I've removed the more intrusive monkey patch in version 3.0.1 of `phipsair` to prevent potential issues with other users of aiocoap. There's one monkey patch left that fixes an exception when shutting down device connections.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.
    - Only partially (config flow). 

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

mfrister committed:
Add scaffolded philips_air_purifier integration

mfrister committed:
Config Flow: Remove not needed credentials

mfrister committed:
Config Flow: Test connection

mfrister committed:
Allow disabling the config entry

mfrister committed:
client: Clean up private variable names

mfrister committed:
Fan: Use purifier icon

mfrister committed:
Use firmware version in addition to wifi firmware version

mfrister committed:
Device Info: Always use data from config entry

We can't query the device on startup and `device_info` is called only
directly after adding the device, so there's no status available yet.

mfrister committed:
ReliableClient: Remove unnecessary duplicate callback

A status callback is enough, there's no need for a separate unavailable
callback - we can just pass None as status.

mfrister committed:
Switch to phipsair and add requirement

mfrister committed:
Client: Use the hass runloop

mfrister committed:
Config Flow: Move connection test to ReliableClient and fix tests

mfrister committed:
Coverage: Exclude untested philips_air_purifier code

elupus commented:
Have you tried to contact the original libs developer to either take of maintainance of his lib or be added as a developer on his lib?

mfrister commented:
@elupus

> Have you tried to contact the original libs developer to either take of maintainance of his lib or be added as a developer on his lib?

No, I hadn't tried to contact the original developer before. There's an [MR on the original repository](https://github.com/betaboon/aioairctrl/pull/5), where different people have mentioned the author over a longer period of time, with no response.

Nevertheless, I've now sent an email to the author, if nothing else, just to let them know what's happening. Thanks for the suggestion!

iMicknl commented on homeassistant/components/philips_air_purifier/__init__.py:
```diff
@@ -0,0 +1,37 @@
+from .client import ReliableClient
+from .const import COAP_PORT, DOMAIN
+
+PLATFORMS: list[str] = ["fan"]
```
```suggestion
PLATFORMS: list[str] = [Platform.FAN]
```

+ `from homeassistant.const import Platform`

iMicknl commented on homeassistant/components/philips_air_purifier/__init__.py:
```diff
@@ -0,0 +1,37 @@
+
+    client = ReliableClient(hass.loop, entry.data["host"], port=COAP_PORT)
+    client.start()
+    if DOMAIN not in hass.data:
+        hass.data[DOMAIN] = {}
+    hass.data[DOMAIN][entry.entry_id] = client
+
```
```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = client
```

iMicknl commented on homeassistant/components/philips_air_purifier/__init__.py:
```diff
@@ -0,0 +1,37 @@
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
+    client = hass.data[DOMAIN][entry.entry_id]
+    client.stop()
```
```suggestion
```

mfrister replied:
I've moved the device_id and device_info to the constructor.

For the name - I think it can change over time, so I'd prefer to use the name received from the device when it's available.

iMicknl commented on homeassistant/components/philips_air_purifier/__init__.py:
```diff
@@ -0,0 +1,37 @@
+    client = hass.data[DOMAIN][entry.entry_id]
+    client.stop()
+
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
```
```suggestion
    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
        client = hass.data[DOMAIN][entry.entry_id]
        client.stop()
        hass.data[DOMAIN].pop(entry.entry_id)
```

mfrister replied:
Nice, I wasn't aware of the `:=` operator 😃

mfrister replied:
[`Status`](https://github.com/mfrister/phipsair/blob/0c148533d88f807a9b7640e7dffe4c04d944bbca/phipsair/purifier.py#L52-L78) contains all information about the state of the purifier, not only information about availability. The `_set_status` method is passed as a callback to the PersistentClient, so it can push status updates to the Fan. I've updated the docstring to make this more clear.

I already had `@property available()` implemented and it indeed seems to stop Home Assistant from calling the other methods, so I could remove the `if`s there (see below).

iMicknl commented on homeassistant/components/philips_air_purifier/__init__.py:
```diff
@@ -0,0 +1,37 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+
```
```suggestion
```

iMicknl commented on homeassistant/components/philips_air_purifier/client.py:
```diff
@@ -0,0 +1,337 @@
+from asyncio.tasks import FIRST_COMPLETED
+from collections.abc import Callable
+from datetime import datetime, timedelta, timezone
+import enum
+from enum import Enum
```
```suggestion
from enum import Enum, unique
```

My preference would be this probably. Or just import enum instead of importing it in two different ways :).

iMicknl commented on homeassistant/components/philips_air_purifier/client.py:
```diff
@@ -0,0 +1,337 @@
+"""client provides a ReliableClient for Philips Air Purifiers."""
```
There is quite some code here that could be moved to the pypi library perhaps? Any reason why you added this to HA core code?

mfrister replied:
You're right, I moved the whole client to the library.

I only discovered quite late that I'd have to publish the library on PyPI myself and didn't rethink this.

iMicknl commented on homeassistant/components/philips_air_purifier/config_flow.py:
```diff
@@ -0,0 +1,67 @@
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required("host"): str,
```
```suggestion
        vol.Required(CONF_HOST): str,
```

iMicknl commented on homeassistant/components/philips_air_purifier/__init__.py:
```diff
@@ -0,0 +1,37 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Philips Air Purifier from a config entry."""
+
+    client = ReliableClient(hass.loop, entry.data["host"], port=COAP_PORT)
```
```suggestion
    client = ReliableClient(hass.loop, entry.data[CONF_HOST], port=COAP_PORT)
```

iMicknl commented on homeassistant/components/philips_air_purifier/config_flow.py:
```diff
@@ -0,0 +1,67 @@
+
+    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
+    """
+    return await ReliableClient.test_connection(data["host"], COAP_PORT)
```
```suggestion
    return await ReliableClient.test_connection(data[CONF_HOST], COAP_PORT)
```

iMicknl commented on homeassistant/components/philips_air_purifier/config_flow.py:
```diff
@@ -0,0 +1,67 @@
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA
+            )
+
```
Instead of twice having the `async_show_form`. You could also just check for `if user_input` and wrap the other part in that.

iMicknl commented on homeassistant/components/philips_air_purifier/config_flow.py:
```diff
@@ -0,0 +1,67 @@
+
+            return self.async_create_entry(
+                title=info["name"],
+                data={"host": user_input["host"], "model": info["model"]},
```
also nice to use const for host / model here

iMicknl commented on homeassistant/components/philips_air_purifier/fan.py:
```diff
@@ -0,0 +1,166 @@
+    """Set up a configured Air Purifier by creating the Fan entity."""
+    client = hass.data[DOMAIN][config_entry.entry_id]
+
+    initial_data = {
+        "unique_id": config_entry.unique_id,
+        "name": config_entry.title,
+        "model": config_entry.data["model"],
+    }
```
Why not just pass the config entry to your entity? I think this will clean up your code.

iMicknl commented on homeassistant/components/philips_air_purifier/fan.py:
```diff
@@ -0,0 +1,166 @@
+class PurifierEntity(FanEntity):
+    """Representation of a Philips air purifier."""
+
+    should_poll: bool = False
```
```suggestion
    _attr_should_poll = False
```

iMicknl commented on homeassistant/components/philips_air_purifier/fan.py:
```diff
@@ -0,0 +1,166 @@
+
+    should_poll: bool = False
+
+    supported_features: int = SUPPORT_PRESET_MODE | SUPPORT_SET_SPEED
```
```suggestion
    _attr_supported_features = SUPPORT_PRESET_MODE | SUPPORT_SET_SPEED
```

iMicknl commented on homeassistant/components/philips_air_purifier/fan.py:
```diff
@@ -0,0 +1,166 @@
+    should_poll: bool = False
+
+    supported_features: int = SUPPORT_PRESET_MODE | SUPPORT_SET_SPEED
+    preset_modes: list[str] = [m.name for m in list(Mode) if m != Mode.Manual]
```
```suggestion
    _attr_preset_modes = [m.name for m in list(Mode) if m != Mode.Manual]
```

iMicknl commented on homeassistant/components/philips_air_purifier/fan.py:
```diff
@@ -0,0 +1,166 @@
+    supported_features: int = SUPPORT_PRESET_MODE | SUPPORT_SET_SPEED
+    preset_modes: list[str] = [m.name for m in list(Mode) if m != Mode.Manual]
+    speeds = [FanSpeed.Speed1, FanSpeed.Speed2, FanSpeed.Speed3, FanSpeed.Turbo]
+    speed_count: int = len(speeds)
```
```suggestion
    _attr_speed_count = len(speeds)
```

iMicknl commented on homeassistant/components/philips_air_purifier/fan.py:
```diff
@@ -0,0 +1,166 @@
+        """Return whether the purifier is available."""
+        return self._status is not None
+
+    @property
+    def icon(self) -> str | None:
+        """Use an air purifier icon instead of the default fan."""
+        return "mdi:air-purifier"
+
```
```suggestion
```
Just set this via `_attr_icon`

iMicknl commented on homeassistant/components/philips_air_purifier/fan.py:
```diff
@@ -0,0 +1,166 @@
+        mode = Mode[preset_mode]
+        await self._client.set_preset_mode(mode)
+
+    device_to_speed_percentage_map = {
+        FanSpeed.Off: 0,
+        FanSpeed.Silent: 10,
+        FanSpeed.Speed1: 25,
+        FanSpeed.Speed2: 50,
+        FanSpeed.Speed3: 75,
+        FanSpeed.Turbo: 100,
+    }
```
Make this a constant in the top of your file.

iMicknl commented on homeassistant/components/philips_air_purifier/manifest.json:
```diff
@@ -0,0 +1,15 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/philips_air_purifier",
+  "requirements": ["phipsair==0.3.1"],
+  "ssdp": [],
+  "zeroconf": [],
+  "homekit": {},
+  "dependencies": [],
```
```suggestion
```

iMicknl commented on homeassistant/components/philips_air_purifier/fan.py:
```diff
@@ -0,0 +1,166 @@
+        """Return the unique ID for this purifier."""
+        if self._status is None:
+            return self._initial_data["unique_id"]
+        return self._status.device_id
```
Can this be moved to constructor? Same as name? Just set the `_attr_name` and unique id there directly.

mfrister replied:
I've moved the `unique_id` to the constructor.

I'd like to keep the name as method, given that the device's name can change over time. This allows showing the current name at least after receiving a status update from the device.

iMicknl commented on homeassistant/components/philips_air_purifier/fan.py:
```diff
@@ -0,0 +1,166 @@
+        self._client.stop_observing_status(id(self))
+
+    def _set_status(self, status: Status | None) -> None:
+        self._status = status
```
What does this Status contain? Why just not set `_attr_available` (or `@property available()`) when the device cannot be accessed? This should also block the other ones from being retrieved.

iMicknl commented on homeassistant/components/philips_air_purifier/fan.py:
```diff
@@ -0,0 +1,166 @@
+        Given that the purifier is sometimes not responsive for 30 seconds ore more,
+        we can't query it during startup and can only use data from the config entry.
+        """
+        info = {
```
Just set `self._attr_device_info` in your constructor. And use the DeviceInfo tuple `from homeassistant.helpers.entity import DeviceInfo`.

iMicknl commented on homeassistant/components/philips_air_purifier/fan.py:
```diff
@@ -0,0 +1,166 @@
+        return info
+
+    @property
+    def is_on(self):
```
Would be nice to try to type your integration fully :). If you want, you can even mark it as strict and fix all mypy errors.

mfrister replied:
Done 🎉 

Fully typed the package phipsair in strict mode as well.

iMicknl commented on homeassistant/components/philips_air_purifier/fan.py:
```diff
@@ -0,0 +1,166 @@
+    @property
+    def percentage(self) -> int | None:
+        """Return the current fan speed converted to a percentage."""
+        if self._status is None:
+            return None
```
I am not a huge fan of this self._status check everywhere. Can you see if this is required when you use `available` (see comment above)?

mfrister replied:
You're right, Home Assistant doesn't seem to call the other methods if the device is unavailable. I've removed the checks everywhere except for the name, where the fallback to the value from the config entry makes sense, I think.

To make typing work, I added assertions, though (`self._status is not None`). Is this the way to go?

iMicknl reviewed (changes requested):
Thanks for your contribution! I had a quick look to get you started :).

mfrister commented:
Thanks for the feedback! I'll have a look and improve things, hopefully in the next few days.

mfrister committed:
Add reliable client and fan

mfrister committed:
Use constants for config entry keys

mfrister committed:
Config Flow: Remove duplicate async_show_form calls

mfrister committed:
Fan: Pass ConfigEntry instead of additional dict

mfrister committed:
Fan: Use HA convention for properties

Also make a device to HA speed map a constant.

mfrister committed:
Use reliable air purifier client from package

Update the package to 0.4.0 with the reliable client.

mfrister committed:
Fan: Use _attr properties for static values

mfrister committed:
Fan: Replace unnecessary status checks with assertions

Home Assistant doesn't seem to call these methods if the device is
unavailable.

mfrister committed:
Fan: Improve name and _set_status docs

mfrister committed:
Enable strict typing incl. phipsair bump to 0.4.1

The new phipsair version has type annotations.

mfrister committed:
Fan: Replace deprecated supported feature constants

mfrister commented:
@iMicknl I think I've addressed all your comments, I added a few comments where things were not completely clear. I think this is ready for another round of review.

I hope it was ok for me to resolve the review conversations that seemed to be clearly resolved - please let me know if you like another workflow better.

Before merging (however far from that we might be 😀), I'd like to have a look at an issue I'm experiencing when running the integration locally in a Docker container, where the device is regularly shown as unavailable when it's turned off.

I suspect this might be an issue with NAT on the machine dropping UDP mappings between status updates from the device. The updates are sent a lot less frequently when the device is turned off (every few minutes instead of every few seconds). The issue never appeared on my development machine where HA is running on the host (it's also a Mac and not a Linux box). I think it would be nice if I could work around that (most certainly in the phipsair package), e.g. by regularly sending packets to the device to keep the NAT mapping alive.

mfrister committed:
Bump phipsair to 0.5.0

Increases reliability when connecting to the purifier via a NAT and
makes the connection test more reliable and faster.

mfrister commented:
@iMicknl I've found a solution to the NAT issues - a regular keepalive request to a simple endpoint ensures the NAT keeps the UDP stream mapping. Device availability in a non-host-networking Docker container has been good since this change.

As a positive side-effect, this has also improved connection test reliability during the config flow - I've discovered a better endpoint while trying to debug the NAT issues.

I made the changes in the phipsair package, I've only added a commit here bumping the version and I've rebased again onto `dev`.

mfrister commented:
@iMicknl @elupus What can I do to help with getting this merged?

mfrister commented:
@kongo09

The custom integration you're maintaining looks really interesting and I think the integration I wrote here could benefit of a lot from parts implemented there. The main reason why I didn't begin with it (or another fork) is the missing license - all the forks from betaboon/denaun don't have a license, so I don't think we can copy anything into Home Assistant core and will basically have to treat the code as proprietary :/ If some parts come from known sources with known licenses (e.g. the icons), we might be able to reuse them if they're compatible with HA's license and the CLA.

Before putting more effort into this, I'd like to see that it's actually viable to get this MR merged - doesn't look easy to get another review so far. Once this is done, let's discuss how we could go forward with extending the core component - there's lots to do. How can I contact you?

kongo09 commented:
I see the problem. I'm happy to contribute whatever I've written myself under whatever license is required, e.g. I've created the icons myself in SVG but the design is obviously done by Philips. Not sure what that means - after all HA is also using brand logos all over the place and I don't think they have a written permission by the brand owners documented anyhwere.

In terms of code, I have massively extended the betaboon stuff, so again, I'm happy to bring that in.

My username here is the same on gmail, so just shoot my a PM.

mfrister commented:
@balloob I've seen you review a few of the recently added integrations and I'm not sure how to continue with this MR after having received no feedback for roughly a month, so I hope you can help me with a bit of advice.

From my point of view, I've addressed all points raised by @iMicknl, either by improving the code or by replying to them. Can I do anything to move this MR forward (does it have the right labels for it to be picked up for another review?) or do I need to adjust my expectations and just wait for a few more weeks?

mfrister commented:
@github-actions I'd still like to get this merged - waiting for a review.

mfrister commented:
@frenck It looks like this PR has been stuck in the review process somehow - I haven’t received any reply since April after putting in considerable effort to address all feedback I previously got.

Could you please help getting this PR going again?

frenck commented:
Start, but I am up in the montains in Austria with my family enjoying my holiday. Please,
In general, don't ping people for review or pull random people into a conversation. 🙏

Thanks!

../Frenck

mfrister commented:
I've now attempted to find someone to give more feedback on this PR over the course of 6 months by different ways of communication, without success.

Unfortunately, I don't have the patience and motivation to continue working on this with no feedback for such a long time, so I will close the PR for now.
